File: python/day_7/part_1.py
def solve(input: str) -> int:
    lines = input.strip().split("\n")
    grid = [list(line) for line in lines]
    rows = len(grid)
    cols = len(grid[0]) if rows > 0 else 0

    start_col = -1
    for col in range(cols):
        if grid[0][col] == "S":
            start_col = col
            break

    beams = {start_col}
    split_count = 0

    for row in range(1, rows):
        new_beams = set()

        for col in beams:
            if 0 <= col < cols:
                if grid[row][col] == "^":
                    split_count += 1
                    if col - 1 >= 0:
                        new_beams.add(col - 1)
                    if col + 1 < cols:
                        new_beams.add(col + 1)
                else:
                    new_beams.add(col)

        beams = new_beams

    return split_count


if __name__ == "__main__":
    with open("inputs/day_7.txt", "r") as f:
        input = f.read()
        result = solve(input)
        print(result)

File: go/day_7/part_1.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func solve(input string) int {
	lines := strings.Split(strings.TrimSpace(input), "\n")
	grid := [][]rune{}
	for _, line := range lines {
		grid = append(grid, []rune(line))
	}
	rows := len(grid)
	cols := 0
	if rows > 0 {
		cols = len(grid[0])
	}

	start_col := -1
	for col := 0; col < cols; col++ {
		if grid[0][col] == 'S' {
			start_col = col
			break
		}
	}

	beams := map[int]bool{start_col: true}
	split_count := 0

	for row := 1; row < rows; row++ {
		new_beams := map[int]bool{}

		for col := range beams {
			if col >= 0 && col < cols {
				if grid[row][col] == '^' {
					split_count++
					if col-1 >= 0 {
						new_beams[col-1] = true
					}
					if col+1 < cols {
						new_beams[col+1] = true
					}
				} else {
					new_beams[col] = true
				}
			}
		}

		beams = new_beams
	}

	return split_count
}

func main() {
	input, err := os.ReadFile("../inputs/day_7.txt")
	if err != nil {
		panic(err)
	}
	result := solve(string(input))
	fmt.Println(result)
}
